### DISCUSSION TITLE
IIListProvider<T> is non-public

### DISCUSSION THREAD

**Discussion by ActivistInvestor** [dotnet/runtime#98886] [Ideas]
How does a user type that implements `IEnumerable<T>` provide a result to TryGetNonEnumeratedCount() ?

It doesn't seem possible. `IIListProvider<T>` is an internal interface. I can't implement it on my type that implements `IEnumerable<T>`, and I'm at a loss to explain why that is.

Any insights on that?

**Comment by huoyaoyuan** [comment]
> How does a user type that implements `IEnumerable<T>` provide a result to TryGetNonEnumeratedCount() ?

You should just implement `ICollection<T>` or `IReadOnlyCollection<T>` and provide `Count`. Having one type changes the ability of getting count is not a good design.

`IIListProvider<T>` is just a LINQ implementation detail. It's used for better optimization of chained LINQ operators.

User code should just implement `IList<T>` or `IReadOnlyList<T>`. `ICollection<T>.CopyTo` can be used to provide some optimization.

**Reply by ActivistInvestor** [reply]
> You should just implement ICollection<T> or IReadOnlyCollection<T> and provide Count. Having one type changes the ability of getting count is not a good design.

I can't implement `ICollection<T>` or `IReadOnlyList<T>`.  The types in question implement only `IEnumerable<T>`, because they cannot always provide a Count, and that isn't known at design-time.

If `TryGetNonEnumeratedCount()` is not applicable to any `IEnumerable<T>,` then it shouldn't be public. It should be an implementation detail also, since it cannot query any user type that implements `IEnumerable<T>`.

**Reply by tannergooding** [reply]
> If TryGetNonEnumeratedCount() is not applicable to any IEnumerable<T>, then it shouldn't be public

The entire point is that its a "try" API, it attempts to get the count. It is a best attempt using the broadly available existing interfaces.

Some new interface could theoretically be provided, but checking if a `T` implements an interface has an actual expense at runtime and adds overhead to every caller. A net new interface that would be implemented by a minority of types that could sometimes, but not always, have a count isn't necessarily worth the additional complexity. It's the same reason why LINQ doesn't check for `IReadOnlyCollection<T>` (a .NET Framework 4.5 interface) in most cases, it only checks for `ICollection<T>` instead

**Reply by huoyaoyuan** [reply]
> The types in question implement only `IEnumerable<T>`, because they cannot always provide a Count, and that isn't known at design-time.

This is the anti-pattern. Collections with different capabilities should be implemented in different types, potentially sharing a common base type.

**Comment by KennethHoff** [comment]
Ironically, this interface was removed mere days after you made this discussion 🤣

https://github.com/dotnet/runtime/pull/98969
